### PR TITLE
Modulable avro plugin

### PR DESCRIPTION
**Pull request by RustedBones**
Currently `sbt-avro` can only use a single avro compiler version in the whole build.

This PR intend to change this by using a custom `avro` configuration, containing the required artifacts for the compilation phase.

Other notable change: As the plugin now adds the `avro` dependency, it is not automatically triggered.

### CODE REVIEW

File: plugin/src/main/scala/com/github/sbt/avro/SbtAvro.scala [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+        "com.github.sbt" % "sbt-avro-compiler-bridge" % BuildInfo.version % Avro,
+        "org.apache.avro" % "avro-compiler" % avroVersion.value % Avro,
+        "org.apache.avro" % "avro" % avroVersion.value,
+      ),
+      libraryDependencies ++= avroAdditionalDependencies.value
+    )
+
+    // settings to be applied for both Compile and Test
+    lazy val configScopedSettings: Seq[Setting[_]] = Seq(
+      avroSources := Seq(sourceDirectory.value / "avro"),
```
Comment (clairemcginty) [inline plugin/src/main/scala/com/github/sbt/avro/SbtAvro.scala]: should we keep the singular `avroSource` setting for backwards compat?

File: plugin/src/main/scala/com/github/sbt/avro/SbtAvro.scala [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+    // format: off
+    val avroAdditionalDependencies = settingKey[Seq[ModuleID]]("Additional dependencies to be added to library dependencies.")
+    val avroCompiler = settingKey[String]("Sbt avro compiler class. Default: com.github.sbt.avro.AvroCompilerBridge")
+    val avroCreateSetters = settingKey[Boolean]("Generate setters. Default: true")
+    val avroDependencyIncludeFilter = settingKey[DependencyFilter]("Filter for including modules containing avro dependencies.")
+    val avroEnableDecimalLogicalType = settingKey[Boolean]("Use java.math.BigDecimal instead of java.nio.ByteBuffer for logical type \"decimal\". Default: true.")
+    val avroFieldVisibility = settingKey[String]("Field visibility for the properties. Possible values: private, public. Default: public.")
+    val avroOptionalGetters = settingKey[Boolean]("Generate getters that return Optional for nullable fields. Default: false.")
+    val avroSpecificRecords = settingKey[Seq[String]]("List of avro records to recompile with current avro version and settings. Classes must be part of the Avro library dependencies.")
```
Comment (clairemcginty) [inline plugin/src/main/scala/com/github/sbt/avro/SbtAvro.scala]:
```suggestion
    val avroSpecificRecords = settingKey[Seq[String]]("List of fully qualified Avro record class names to recompile with current avro version and settings. Classes must be part of the Avro library dependencies. Default: Nil")
```

File: plugin/src/main/scala/com/github/sbt/avro/SbtAvro.scala [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    val externalSrcDir = (avroUnpackDependencies / target).value
+    val srcDirs = Seq(externalSrcDir) ++ avroSources.value
+    val outDir = (key / target).value
+
+    val cachedCompile = {
+      import sbt.util.CacheStoreFactory
+      import sbt.util.CacheImplicits._
+
+      val cacheStoreFactory = CacheStoreFactory(out.cacheDirectory / "avro")
+      val lastCache = { (action: Option[Set[File]] => Set[File]) =>
```
Comment (clairemcginty) [inline plugin/src/main/scala/com/github/sbt/avro/SbtAvro.scala:193]: Do we need to include Avro compiler version in the cache key as well?

File: plugin/src/main/scala/com/github/sbt/avro/SbtAvro.scala (hunk 1 above)
Comment (RustedBones) [reply plugin/src/main/scala/com/github/sbt/avro/SbtAvro.scala]: Yes. I'll add the `avroUmanagedSourceDirectories` instead to match with default sbt behavior

File: plugin/src/main/scala/com/github/sbt/avro/SbtAvro.scala (hunk 3 above)
Comment (RustedBones) [reply plugin/src/main/scala/com/github/sbt/avro/SbtAvro.scala:193]: changing any sbt setting involves to reload the project to take effect. the reloading operation invalidates all caches.

File: bridge/src/main/java/com/github/sbt/avro/AvroCompilerBridge.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            System.out.println("Compiling Avro schema: " + ref.getFile());
+            files.add(ref);
+        }
+        compiler.compileFiles(Set.of(avscs), target);
+    }
+
+    @Override
+    public void compileAvprs(File[] avprs, File target) throws Exception {
+        for (File avpr : avprs) {
+            System.out.println("Compiling Avro protocol " + avpr);
```
Comment (kellen) [inline bridge/src/main/java/com/github/sbt/avro/AvroCompilerBridge.java]:
```suggestion
            System.out.println("Compiling Avro protocol: " + avpr);
```

File: bridge/src/main/java/com/github/sbt/avro/AvroCompilerBridge.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        compiler.setFieldVisibility(fieldVisibility);
+        compiler.setUseNamespace(useNamespace);
+        compiler.setEnableDecimalLogicalType(enableDecimalLogicalType);
+        compiler.setCreateSetters(createSetters);
+        if (AvroVersion.getMinor() > 8) {
+            compiler.setGettersReturnOptional(optionalGetters);
+        }
+        if (AvroVersion.getMinor() > 9) {
+            compiler.setOptionalGettersForNullableFieldsOnly(optionalGetters);
+        }
```
Comment (kellen) [inline bridge/src/main/java/com/github/sbt/avro/AvroCompilerBridge.java]: These chunks seem repeated; are they not possible to extract?

File: bridge/src/main/java/com/github/sbt/avro/AvroCompilerBridge.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+            compiler.setStringType(stringType);
+            compiler.setFieldVisibility(fieldVisibility);
+            compiler.setEnableDecimalLogicalType(enableDecimalLogicalType);
+            compiler.setCreateSetters(createSetters);
+            if (AvroVersion.getMinor() > 8) {
+                compiler.setGettersReturnOptional(optionalGetters);
+            }
+            if (AvroVersion.getMinor() > 9) {
+                compiler.setOptionalGettersForNullableFieldsOnly(optionalGetters);
+            }
```
Comment (kellen) [inline bridge/src/main/java/com/github/sbt/avro/AvroCompilerBridge.java]: Shouldn't the AvroVersion.getMinor ifs also check major version?

Also I suppose maintaining this bridge class means we need to track the avro API changes with each new version to correctly pass the new parameters. Is that right?

File: plugin/src/main/scala/com/github/sbt/avro/SbtAvro.scala [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+import sbt.Keys.*
+import sbt.{Def, *}
+import Path.relativeTo
+import sbt.librarymanagement.DependencyFilter
+
+import java.io.File
+import java.net.URLClassLoader
+
+/**
+ * Simple plugin for generating the Java sources for Avro schemas and protocols.
```
Comment (kellen) [inline plugin/src/main/scala/com/github/sbt/avro/SbtAvro.scala]:
```suggestion
 * Plugin for generating the Java sources for Avro schemas and protocols.
```

File: bridge/src/main/java/com/github/sbt/avro/AvroCompilerBridge.java (hunk 5 above)
Comment (RustedBones) [reply bridge/src/main/java/com/github/sbt/avro/AvroCompilerBridge.java]: Unfortunately the compilers do not share a common interface

File: bridge/src/main/java/com/github/sbt/avro/AvroCompilerBridge.java (hunk 6 above)
Comment (RustedBones) [reply bridge/src/main/java/com/github/sbt/avro/AvroCompilerBridge.java]: yes. If there is a new compiler parameter, we can add it to the sbt plugin but we must make sure that if the runtime avro version does not support it, we won't call the method.

File: plugin/src/main/scala/com/github/sbt/avro/SbtAvro.scala (hunk 7 above)
Comment (RustedBones) [reply plugin/src/main/scala/com/github/sbt/avro/SbtAvro.scala]: Yes! :laughing:

File: bridge/src/main/java/com/github/sbt/avro/AvroCompilerBridge.java (hunk 5 above)
Comment (kellen) [reply bridge/src/main/java/com/github/sbt/avro/AvroCompilerBridge.java]: Ahh, fun

**Pull request merged**